### ISSUE TITLE
Support for EQ mounts

### ISSUE THREAD

**Issue by Serphentas** [nforrester/NexPlane#1]
First of all, thanks for writing this software. It's an amazing jumpstart to track fast objects, especially the ISS.

I think it would be really nice to have support for EQ mounts (CGx or wedge) as it avoids removing converting the setup each time, granted it is possible (wedge users only).

I see that [these modes](https://github.com/nforrester/NexPlane/blob/25e86a7be8c6886daf554928fc5d86f10ffd49a2/nexstar.py#L78) are hard disabled. My guess is that the calculations are very different and you've not had time to implement those yet.

Could you give points on this matter ? I would gladly help in implementing this feature.

**Comment by nforrester** [comment]
Thanks, I'd appreciate the help!

You're not the first person to request support for equatorial mounts. The only reason I haven't done it already is that the only mount I have that's currently compatible with NexPlane is Alt/Az (NexStar 8SE). I'd hoped when I bought my SkyWatcher AZ-EQ6 that it would just work out of the box, but unfortunately the firmware on that mount is not capable of fast variable-rate slewing (it can do fast fixed-rate slewing, or slow variable-rate slewing, but not both). I asked SkyWatcher for the firmware source code so I could help them fix it, but they declined. As such, my next big astronomy project is to build and program a new motherboard for it (which I'll happily open source when I'm done). After that I had a vague idea that I might make NexPlane work with EQ mounts (since I'd then have a compatible mount that could be EQ). It could take me three years to get to that point though, since all of this is in my free time.

Regarding the TrackingMode enum class you found, you can go ahead and enable those modes. However, that will not help make NexPlane compatible with EQ mounts. The only use for the TrackingMode class is to help NexPlane set the mount's tracking mode to OFF (this class is how NexPlane knows what numerical value in the serial interface means OFF). NexPlane doesn't want the mount trying to track celestial bodies independently, NexPlane wants to directly control where the mount points relative to the surface of the Earth, and if it wants the mount to move at a certain rate it will just command that slew rate directly.

Here's the general approach I would suggest you take:

Currently NexPlane computes the Az/El position of the target and sends it to the Tracker object [here](https://github.com/nforrester/NexPlane/blob/a402635e23b46ba2463c6c5200ebbc31f0777abb/nexplane.py#L179). The Tracker object runs a PID controller for each of the Az and El axes and sends slew rate commands to the NexStar object [here](https://github.com/nforrester/NexPlane/blob/4844e9f80ad8081cea1fc4326c52a54ba8e936a8/tracker.py#L95). The NexStar object accepts those slew rates and transmits them to the mount over the serial port [here](https://github.com/nforrester/NexPlane/blob/4844e9f80ad8081cea1fc4326c52a54ba8e936a8/nexstar.py#L592).

If you have a NexStar mount like mine, but you have a wedge for it, I suggest that you create new functions for the NexStar class that command Right Ascension and Declination slew rates. The slew_ra() function can just delegate directly to slew_azm() and slew_dec() can delegate to slew_alt(), because of course that's what those axes become when you put the mount on a wedge (and the [documentation](https://www.nexstarsite.com/PCControl/ProgrammingNexStar.htm) explicitly calls out the relevant serial commands as being dual purpose).

If you have some other mount, I suggest you create a new class, similar to NexStar, that performs similar functions for your mount.

Once you have a class that lets you slew your mount along RA and Dec, I suggest that you add a function that converts the target coordinates from Az/El to RA/Dec, and then use the Tracker class to make your new mount interface class track the target along these axes.

Once you've got all that working, you can add [config](https://github.com/nforrester/NexPlane/blob/master/config_default.yaml) and [command line](https://github.com/nforrester/NexPlane/blob/a402635e23b46ba2463c6c5200ebbc31f0777abb/nexplane.py#L28) options to choose a mount type.

Lastly, you may wish to add a [marker on the GUI](https://github.com/nforrester/NexPlane/blob/a402635e23b46ba2463c6c5200ebbc31f0777abb/gui.py#L194) to indicate the north pole in the sky. On an Alt/Az mount there's a singularity near the vertical where the mount experiences gimbal lock and is almost guaranteed to lose tracking. The GUI conveniently places this at the top of the screen. However, on an equatorial mount that singularity is located at the north pole, so it would helpful to indicate on the screen where this is so that the user will be aware when a target is flying towards it.

Please share your changes, and feel free to ask more questions. Good luck!

**Comment by Serphentas** [comment]
Thank you very much for this comprehensive post ! I will have some free time in August, so we can expect a PR around that timeframe.

**Comment by Serphentas** [comment]
I've implemented most features except the AltAz->RaDec transformation. There is a fancy [transform_to](https://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.transform_to) method to go from one frame to another in AstroPy ([example](https://gist.github.com/dokeeffe/18857db66dbabc14679c20a8560e2cd6)) but it seems the time must be be set each time. I also found [this](https://jonvoisey.net/blog/2018/07/data-converting-alt-az-to-ra-dec-derivation/) blog post, not quite sure which way to go from here.

Any suggestions ?

**Comment by nforrester** [comment]
Yes, this is a little tricky. RA/Dec is an inertial frame fixed to the stars, whereas Alt/Az is a rotating frame fixed to the Earth. In order to transform from one to the other, you need to know the angle of the Earth relative to the stars. Astropy can figure this out if it knows the date and time, because fortunately the Earth's rotation is very predictable.

The simplest approach is to do a polar alignment on your mount before you start nexplane so that the mount's internal software understands the relationship between the Earth, the stars, and its motors. Then you just tell astropy what time it is using the python datetime library and you're good to go. This assumes that you can see enough stars to do a polar alignment, which may not be true during the day. Despite this limitation, **I recommend that you use this approach.**

If you want to be able to use nexplane with an equatorial mount that has _not_ been polar aligned, things get more complex. I had a similar wish for my alt/az mount. I can orient it vertically very easily, but aligning it to the sun or stars is not always possible. For example, if I wish to set up indoors and look out the window at airplanes (a very useful test environment), and the sun is not shining through the window directly on the telescope, aligning on the sun is not possible. To solve this problem I added a ["landmark alignment" mode](https://github.com/nforrester/NexPlane/blob/a402635e23b46ba2463c6c5200ebbc31f0777abb/nexplane.py#L126). I configured the latitude, longitude, and elevation of a distant object on the surface of the Earth that I could reliably see through my window. Then before starting nexplane I pointed my telescope at that object, and then told nexplane that that's where the telescope started. Now, even though the mount itself does not know where it is pointed in the sky, nexplane knows the calibration between the real alt/az coordinate system and the mount's unaligned coordinate system, and can compensate using the `azm_cal` and `alt_cal` variables. You could, with enough effort, create a similar system for an equatorial mount. However, because you cannot be confident that the RA axis of your mount is actually aligned with the rotational axis of the Earth, you would need additional calibration landmarks and more complex math to perform this calibration. Also the built in astropy functions may not be designed with this use case in mind, so you would need to get creative about feeding them the right lies to produce the transform you need. If you want to go down this rabbit hole, we should probably discuss it by phone. Send me an email and we can set something up.

**Comment by Serphentas** [comment]
Thanks again for the comprehensive reply.

The native alignment is clearly the way to go, saving us a ton of headache. There [are](https://play.google.com/store/search?q=polar+align&c=apps) [apps](http://astro.ecuadors.net/polar-scope-align/) that allow to bypass stars in daytime, which could be a viable option whose accuracy has to be tested.

I will go down that path and let you know, but the landmark way can be kept in a box... for later.

**Comment by 1968442** [comment]
Neil.  Having to do with this subject, have you thought about a GOTO telescope object (plane) tracker that instead of using preordained coordinates, uses live video feedback to lock and slew with that object (while also having a safety zone so telescope will stop if too close to the sun like your current program)?  There was a software object tracking program (OpticTracker.com), while that website is still there, the company is not and even their Facebook page hasn’t been updated past 2013ish…
 You seem to have a knowledge base suited to such a task.

**Comment by nforrester** [comment]
That would be very cool! I'm not experienced with live video processing like that, but maybe someday I will have the necessary time to devote to it.

If you have the necessary time, open a new issue thread and we can get into specifics about how your feature could integrate with the existing system.

**Comment by 1968442** [comment]
Thank you Neal.

**Comment by Serphentas** [comment]
To continue on the topic of video guiding, I've seen a demo of such programs and it's really nice as you can avoid theoretical guiding that's often wrong and switch to accurate commands that move just enough to track well.

My experience in the current state (AltAz + Solar alignment) is that most of the time it goes to where the plane roughly is, so you still have to look around with the 9x50 finderscope. After you've found it there and centered it in the ocular/camera, it still has trouble staying on the target.

Given this, I think the error introduced during alignment and the theoretical aspect of coordinates-based guiding can only get us so far. Perhaps a two-headed system could help: theoretical guiding to get near the plane, and on camera capture perform a soft handover to video guiding so that it corrects the mount just enough and where it's really needed.

I do have experience with this and ML, and the couple videos I've capture recently could serve as datasets for training.

What do you think ?

**Comment by 1968442** [comment]
So here is a question that may also be an ethics question…  There is a dead website Optictracker, that the owner seems to have gone out of business about 10 years ago.  The website used to sell this very software to allow a goto telescope to track objects using a digital camera.  The website is still there and they still have the software as a download on that site that would work with a plethora of telescope platforms.  They not only left the software download on that site, but also a downloadable instruction manual.  The only issue (besides that they won’t respond to give you an email key to use it to run the executable file with the key to launch it, I tried to setup an account through their website just in case and their email comes back as undeliverable), is, (would it be ethical if the owner could not be found to ask permission), find a way to bypass the executable file to use the software that is still there, or possibly open the coding that is used to track and post code to specific telescopes?  Is a defunct company’s (by 10 years) downloadable program something that is allowed to be studied and deciphered/key bypassed?

**Comment by nforrester** [comment]
> Is a defunct company’s (by 10 years) downloadable program (that is still downloadable) something that is allowed to be studied and deciphered/key bypassed?

I'm no lawyer, but my understanding of US copyright law suggests that this would not be legal. You would need to get permission from the copyright holder, and if you can't find them then your only legal recourse is to wait many decades for the work to pass into the public domain.

I can't tell you what to do in the privacy of your own computer, but I will not accept PRs into NexPlane that encourage cracking proprietary software.

**Comment by nforrester** [comment]
> My experience in the current state (AltAz + Solar alignment) is that most of the time it goes to where the plane roughly is, so you still have to look around with the 9x50 finderscope. After you've found it there and centered it in the ocular/camera, it still has trouble staying on the target.

Yes, this is about how well it works for me too. It's pretty good, but definitely needs manual corrections.

> Given this, I think the error introduced during alignment and the theoretical aspect of coordinates-based guiding can only get us so far. Perhaps a two-headed system could help: theoretical guiding to get near the plane, and on camera capture perform a soft handover to video guiding so that it corrects the mount just enough and where it's really needed.
> 
> I do have experience with this and ML, and the couple videos I've capture recently could serve as datasets for training.
> 
> What do you think ?

Sounds great to me.

The simplest interface between the video processor and the existing NexPlane software might be similar to the current keyboard interface: "go up!" "go right!". The video processor could also emit axis speed commands directly.

I think it would be challenging for the video processor to accurately emit commands of the form "the target is at X azimuth and Y elevation, so look there!". This would probably just amplify the current calibration inaccuracies and might result in an unstable feedback loop. If you want to try it though you're welcome to do so.

I would like to preserve the option to operate NexPlane manually, for those who don't have a camera. I would also like clear documentation about what kind of camera, cables, etc one would need. I'd be happy to help wordsmith this if you can provide the raw information (I don't know much about this subject).

I look forward to seeing what you come up with!

**Comment by Serphentas** [comment]
> So here is a question that may also be an ethics question [...]

As Neil said, anything can be done in private but even that is, I think, far too much work than reinventing the wheel with current day software and tools. We've come a long way since, ML is very easy to use for such tasks, and it would be simply faster to implement natively in Python than try to understand and convert old code.

> The simplest interface between the video processor and the existing NexPlane software might be similar to the current keyboard interface: "go up!" "go right!". The video processor could also emit axis speed commands directly.

> I think it would be challenging for the video processor to accurately emit commands of the form "the target is at X azimuth and Y elevation, so look there!". This would probably just amplify the current calibration inaccuracies and might result in an unstable feedback loop. If you want to try it though you're welcome to do so.

Yes, I think it's best to keep the current flow of information/commands and implement video guiding as yet another source of command, not a bypass. It's more elegant and should keep most of the current codebase intact.

> I would like to preserve the option to operate NexPlane manually, for those who don't have a camera. I would also like clear documentation about what kind of camera, cables, etc one would need. I'd be happy to help wordsmith this if you can provide the raw information (I don't know much about this subject).

On this topic, I've been working on a PR to manually control the mount using WASD when it is not tracking. It works great with an Xbox controller that is mapped to a bunch of keys (WASD being some of them). This avoid the need to steer out of the Sun with the hand controller. If time permits, I'll publish it this weekend.

As for the camera and documentation clarification, it's much needed indeed. There are a million ways to do video, so perhaps drafting a "specification" or "video process" will help in that regard. For instance, let's imagine:

`[camera library] -> [video library] -> [NexPlane]`

Here, the video library is agnostic to the actual hardware used for capture (think Video4Linux) which in turn allows us to add support for cameras as time passes. ASI would be a great start.

**Comment by nforrester** [comment]
That sounds logical.

**Comment by 1968442** [comment]
> > Is a defunct company’s (by 10 years) downloadable program (that is still downloadable) something that is allowed to be studied and deciphered/key bypassed?
> 
> I'm no lawyer, but my understanding of US copyright law suggests that this would not be legal. You would need to get permission from the copyright holder, and if you can't find them then your only legal recourse is to wait many decades for the work to pass into the public domain.
> 
> I can't tell you what to do in the privacy of your own computer, but I will not accept PRs into NexPlane that encourage cracking proprietary software.

That sounds completely fair and I agree!  I can delete my question if you feel it is encouraging any future reader to be dishonest.  My apologies.

**Comment by nforrester** [comment]
I see no need to delete the discussion. Wondering what might be legal or ethical is perfectly legal and ethical. Having completed that discussion, we can move on.

**Comment by Thomas-NY** [comment]
Since nobody's mentioned yet, I'd like to note that implementing something like this is treading a fine line-Tracking aircraft with ADSB and visually then issuing corrective commands based on alt-az comes somewhat close to what things like targeting pods and missiles do- and could be considered an [ITAR ](https://en.wikipedia.org/wiki/International_Traffic_in_Arms_Regulations)issue. (Specifically section XII)

Obviously this isn't military software, and as-is isn't capable of doing any harm (unless you attached a laser pointer to the telescope or something dumb), but the government would probably prefer being overzealous with its coverage versus missing something. If people in the model rocket hobby can't speak about guidance without concern, it's probably safe to assume we're treading a fine line here. 

This is somewhat discusses in [this](https://www.cloudynights.com/topic/625961-go-to-planespotting-and-motion-tracking/) thread (this thread is how I found NexPlane). The "specifically designed" clause is a serious consideration here, as a version of NexPlane with optical tracking would really push a telescope from "enhanced star viewing device" to "aiming-at-aircraft-accurately device".

By no means am I saying this project should be discontinued-This software is amazing. But I think it's something that needs serious consideration before proceeding to come up with a international-friendly solution.

**Comment by nforrester** [comment]
It's a good question. I'll peruse section XII in more detail this weekend. It and the other sections are available here: https://www.ecfr.gov/current/title-22/chapter-I/subchapter-M/part-121. But I took a quick glance at it.

My first instinct is to say that telescope mounts and weapon mounts are unlikely to have compatible software interfaces. Telescope mounts are also quite small and not resistant to bumps and jolts, as a mount for non-laser weapons must be, and are therefore a poor choice in a military application. As long as we don't write a control interface for a weapon mount then the software is not useful for non-laser weapons.

Furthermore, the software is not capable of planning an intercept between two flying objects. In order to hit anything with a projectile you need to lead the target by a calculated amount, and even a video-augmented-NexPlane provides no assistance with this task.

Obviously shining lasers at airplanes is illegal, and people get arrested for that all the time without the aid of telescopes or special software. ITAR section XII does mention lasers, but only those related to the following purposes:

- Target designators
- Range finders
- Sensors designed to detect retroreflective objects
- LIDAR
- Lasers funded by the DoD
- Lasers that mount on weapons
- Lasers of various other descriptions

The only part of this that a video-augmented-NexPlane could conceivably be useful for would be to help shine a laser on a moving object, which could be useful as a target designator. It wouldn't completely solve the problem though, and it wouldn't be "specially designed" for the purpose since we're designing this to make cool videos, not aim lasers.

My guess is we'd be in the clear, but I'll read more carefully this weekend.

**Comment by Serphentas** [comment]
That's an interesting point being raised by Thomas. Myself being outside the EU (Switzerland), let alone the US, the impact of ITAR is negligible I think. However, one could argue that this software was originally written in the US and/or is hosted on GitHub which is American, and as such can be subjected to restrictions.

Nevertheless, ADS-B only allows us to track transponders using civilian modes (A/C/S) as used by civilian aircrafts at all times and military aircrafts at select times, while military traffic ought to be using military modes (1, 2, 4, 5) the rest of the time.

![image](https://github.com/nforrester/NexPlane/assets/9091195/c438e566-0001-4195-b1ea-ee62d99ac64f)
 
Therefore, the military itself isn't threatened by any means. It only concerns civilian traffic, the impact of which may be overstated. Telescope mounts, in our case NexStar, aren't really fast, and will have difficulty tracking close targets in the hypothetical scenario of nefarious laser disturbance. Likewise, further planes can be tracked well but the laser beam will quickly lose its strength, rendering the intended use void. This balance is difficult to measure or quantify, so checking US laws in parallel doesn't hurt.

**Comment by nforrester** [comment]
I reviewed ITAR section XII in detail. Only four items seem potentially relevant, and I believe none of them really apply, but I am not a lawyer:

> (a) Fire control, aiming, detection, guidance, and tracking systems, as follows:
> [...]
> * (7) Missile or ordnance electro-optical tracking systems;

You could theoretically use a video-augmented-NexPlane to track missiles or ordnance, if you could get it pointed at such an object in the first place. NexPlane would provide no help at all in such initial pointing. Moreover, I _think_ (though it's ambiguous) that this is referring to electro-optical tracking systems designed to be _mounted on_ missiles or ordnance, which NexPlane certainly is not (it assumes a fixed location on the Earth).

> * (b) Laser systems and end items, as follows:
> (1) Laser target designators or coded target markers, that mediate the delivery of ordnance to a target;

You could maybe use a video-augmented-NexPlane as part of a crappy target designator, but it wouldn't help at all with aligning the laser with the camera. NexPlane certainly is not designed for such use. Also you could maybe use off the shelf laser diodes as part of a crappy target designator, but those diodes aren't ITAR.

> * (c) Imaging systems or end items, as follows:
> [...]
> (3) Electro-optical reconnaissance, surveillance, target detection, or target acquisition systems, specially designed for articles in this subchapter or specially designed for a military end user (MT if for determining bearings to specific electromagnetic sources (direction finding equipment) or terrain characteristics and designed or modified for rockets, missiles, SLVs, drones, or unmanned aerial vehicle systems capable of delivering at least a 500 kg payload to a range of at least 300 km);

A video-augmented-NexPlane would arguably be part of an "Electro-optical reconnaissance, surveillance, target detection, or target acquisition system", but it would certainly not be "specially designed for articles in this subchapter or specially designed for a military end user", so this sentence wouldn't seem to apply. The parenthetical about the MTCR also doesn't seem to apply.

> (4) Infrared search and track (IRST) systems having one of the following:
> (i) Airborne or naval systems, that:
> (A) Have range performance of 3 km or greater;
> (B) Incorporate or are specially designed to incorporate an infrared focal plane array or imaging camera, having a peak response wavelength exceeding 3 microns or greater; and
> (C) Maintain positional or angular state of a target through time; or

NexPlane will not work when used on a plane, and you'd have to jump through hoops to use it on a boat (i.e. make sure the boat is stationary and highly stable). Furthermore even a video-augmented-NexPlane will not help at all with a "search" task, only with a track task, because you need to put the object of interest in frame yourself. Also it certainly wouldn't incorporate or be specially designed for an infrared camera with any specific characteristics. Most users will be using visible-light cameras.

I also reviewed EAR category 6 about sensors and lasers, but it seems intensely focused on very specific physical devices.

Summary: I think we're in the clear, but, and I cannot stress this enough: I am not a lawyer. Anybody who isn't comfortable contributing to a video-augmentation project for NexPlane is more than welcome to not contribute. I'm comfortable though.

**Comment by nforrester** [comment]
Well I just ordered a wedge for my NexStar mount and I'd like to use it with NexPlane next week. Have you made any more progress with equatorial mount support, @Serphentas? I'd like to take a crack at it this weekend, but I don't want to duplicate effort you've already made.

**Comment by nforrester** [comment]
I finished adding support for equatorial mounts, @Serphentas. I haven't tested it outside yet, so I haven't merged to master. I'll test it this week and merge next weekend. Feel free to take a look and try it yourself if you dare. :-)

Here's the pull request: https://github.com/nforrester/NexPlane/pull/4

**Comment by nforrester** [comment]
It works great. Merging, and closing this issue.

**Comment by nforrester** [comment]
Functional test data proving it works. :-D

https://www.youtube.com/watch?v=oEUCSdi0gBk

If anybody wants to further discuss closed loop video augmentation, feel free to open a new Issue about it.
